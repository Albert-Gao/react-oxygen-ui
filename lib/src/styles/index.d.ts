import { Responsive } from '../components';
import { defaultTheme } from './theme/theme';
export declare type ThemeType = typeof defaultTheme;
export declare type SpaceType = Responsive;
export declare type FontSizeType = Responsive;
export declare type GradientType = keyof ThemeType['gradients'];
export declare type ColorType = string | string[];
export declare type FontWeightType = keyof ThemeType['fontWeights'];
export declare type BoxShadowType = number | number[];
export { DEFAULT_FONT_SIZE, fontSpecs } from './theme/constants/fontSpecs';
export { DEFAULT_SPACE } from './theme/constants/spaces';
export { DEFAULT_SIZE } from './theme/constants/sizes';
export { DEFAULT_COLORS } from './theme/constants/colors';
export { defaultTheme } from './theme/theme';
export * from './theme/theme.utils';
export { styled, css, createGlobalStyle, keyframes, ThemeProvider, } from './styled';
export { getTransitionValue, getSafeValue, getSafeCss, } from './utils';
