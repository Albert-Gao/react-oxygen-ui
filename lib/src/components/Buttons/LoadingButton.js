"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const styles_1 = require("../../styles");
const utils_1 = require("../../utils/utils");
const BaseBox_1 = require("../Box/BaseBox");
const Image_1 = require("../Image");
const shared_1 = require("./shared");
const threeDots_svg_1 = __importDefault(require("./threeDots.svg"));
exports.LoadingButton = ({ isLoading, noDefaultHoverCSS, children, css, takeRef, ...rest }) => {
    const cssToAdd = `
    ${isLoading ? 'pointer-events: none' : ''};
    ${shared_1.getBaseButtonCSS(noDefaultHoverCSS)};
    ${shared_1.baseNonFlatButtonCSS};
    ${styles_1.getSafeCss(css)};
  `;
    const a11y = {
        'aria-hidden': 'true',
    };
    return (<BaseBox_1.BaseBox as="button" position="relative" border={0} borderColor="transparent" borderRadius={0} fontSize="bodyHeading" px={styles_1.DEFAULT_SPACE.small} minHeight="44px" maxHeight="44px" bg="primary" color="white" css={cssToAdd} {...utils_1.getRefProp(takeRef)} {...rest}>
      <Image_1.Image position="absolute" left="50%" top="50%" src={threeDots_svg_1.default} alt="loading spinner" height="8px" opacity={isLoading ? 1 : 0} {...a11y} transform={isLoading
        ? 'translateX(-50%) translateY(-50%) scale(1)'
        : 'translateX(-50%) translateY(-50%) scale(3)'}/>
      <BaseBox_1.BaseBox opacity={isLoading ? 0 : 1} transform={isLoading ? 'scale(0.3)' : 'scale(1)'}>
        {children}
      </BaseBox_1.BaseBox>
    </BaseBox_1.BaseBox>);
};
