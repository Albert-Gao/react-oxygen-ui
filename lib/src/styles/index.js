"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var fontSpecs_1 = require("./theme/constants/fontSpecs");
exports.DEFAULT_FONT_SIZE = fontSpecs_1.DEFAULT_FONT_SIZE;
exports.fontSpecs = fontSpecs_1.fontSpecs;
var spaces_1 = require("./theme/constants/spaces");
exports.DEFAULT_SPACE = spaces_1.DEFAULT_SPACE;
var sizes_1 = require("./theme/constants/sizes");
exports.DEFAULT_SIZE = sizes_1.DEFAULT_SIZE;
var colors_1 = require("./theme/constants/colors");
exports.DEFAULT_COLORS = colors_1.DEFAULT_COLORS;
var theme_1 = require("./theme/theme");
exports.defaultTheme = theme_1.defaultTheme;
__export(require("./theme/theme.utils"));
var styled_1 = require("./styled");
exports.styled = styled_1.styled;
exports.css = styled_1.css;
exports.createGlobalStyle = styled_1.createGlobalStyle;
exports.keyframes = styled_1.keyframes;
exports.ThemeProvider = styled_1.ThemeProvider;
var utils_1 = require("./utils");
exports.getTransitionValue = utils_1.getTransitionValue;
exports.getSafeValue = utils_1.getSafeValue;
exports.getSafeCss = utils_1.getSafeCss;
