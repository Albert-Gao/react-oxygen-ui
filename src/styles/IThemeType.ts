//
// The following are the expected properties for your styled-components' theme object
// styled-system depends on them to provide you all the conveniences
//
// [Key]	        [Type]	          [Description]
// breakpoints	  Array	            Array of viewport widths to use for min-width media queries
// fontSizes	    Array	            Array of numbers to use as a typographic scale
// colors	        Object	          Color names that can be used in color, bg, and borderColor props
// space	        Array	            Array of numbers for use as margin and pixel values
// fonts	        Array or Object	  Values for the fontFamily prop
// fontWeights	  Array or Object	  Values for fontWeight prop
// lineHeights	  Array or Object	  Values for lineHeight prop
// letterSpacings	Array or Object	  Values for letterSpacing prop
// shadows	      Array or Object	  Values for boxShadow prop
// borders	      Array or Object	  Values for border props
// radii	        Array or Object	  Values for borderRadius props
// opacity	      Array or Object	  Values for opacity props

type NumberOrString = number | string;

type ObjectOrArray<T> =
  | {
      [propName: string]: T;
    }
  | T[];

export interface ITheme {
  breakpoints: NumberOrString[];
  fontSizes: NumberOrString[];
  colors: { [colorName: string]: string };
  space: NumberOrString[];
  fontWeights: ObjectOrArray<number>;
  lineHeights?: ObjectOrArray<NumberOrString>;
  borders: ObjectOrArray<NumberOrString>;
  radii: ObjectOrArray<string>;
  shadows: ObjectOrArray<string>;
  opacity: ObjectOrArray<string>;
  gradients?: object;
}

export const defaultTheme: ITheme = {
  breakpoints: ['40em', '52em', '64em'],
  fontSizes: ['12px', '14px', '16px', '24px', '30px', '36px', '48px'],
  colors: {
    white: 'white',
  },
  fontWeights: [100, 300, 400, 500, 700, 900],
  borders: [0, '1px solid'],
  radii: ['4px'],
  shadows: [
    '0px 1px 2px rgba(0,0,0,0.20)',
    '0px 2px 4px rgba(0,0,0,0.20)',
    '0px 4px 8px rgba(0,0,0,0.20)',
    '0px 8px 16px rgba(0,0,0,0.20)',
    '0px 12px 24px rgba(0,0,0,0.20)',
  ],
  opacity: [],
  space: [0, 4, 8, 16, 24, 32, 48, 64, 96, 128, 256, 512],
};
