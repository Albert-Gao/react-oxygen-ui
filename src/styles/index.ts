import { Responsive } from '../components';
import { defaultTheme } from './theme/theme';


export type ThemeType = typeof defaultTheme;
export type SpaceType = Responsive;
export type FontSizeType = Responsive;
export type GradientType = keyof ThemeType['gradients'];
export type ColorType = string | string[];
export type FontWeightType = keyof ThemeType['fontWeights'];
export type BoxShadowType = number | number[];

export { DEFAULT_FONT_SIZE, fontSpecs } from './theme/constants/fontSpecs';
export { DEFAULT_SPACE } from './theme/constants/spaces';
export { DEFAULT_SIZE } from './theme/constants/sizes';
export { DEFAULT_COLORS } from './theme/constants/colors';
export { defaultTheme } from './theme/theme';
export * from './theme/theme.utils';

export {
  styled,
  css,
  createGlobalStyle,
  keyframes,
  ThemeProvider,
} from './styled';

export {
  getTransitionValue,
  getSafeValue,
  getSafeCss,
} from './utils';
