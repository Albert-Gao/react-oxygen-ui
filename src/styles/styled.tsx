import React from 'react';
import * as styledComponents from 'styled-components';
// tslint:disable-next-line:no-duplicate-imports
import { ThemedStyledComponentsModule } from 'styled-components';
import { ThemeType } from './index';
import { defaultTheme, ITheme } from './theme/theme';

const ThemeProvider: React.FC<{
  theme: ITheme;
  // tslint:disable-next-line:no-any
  children: React.ReactElement<any>;
}> = ({ theme, children }) => (
  <styledComponents.ThemeProvider theme={theme || defaultTheme}>
    {children}
  </styledComponents.ThemeProvider>
);

const {
  default: styled,
  css,
  createGlobalStyle,
  keyframes,
} = styledComponents as ThemedStyledComponentsModule<ThemeType>;

export { styled, css, createGlobalStyle, keyframes, ThemeProvider };
