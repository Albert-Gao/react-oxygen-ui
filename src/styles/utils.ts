import { FlattenInterpolation } from 'styled-components';
import { createMediaQuery } from 'styled-system';
import { CSSType } from '../components/Box/BaseBox.type';
import { is } from '../utils/utils';
import { ITheme } from './IThemeType';

export const getMediaQuery = (
  screenSize: number | string,
  css: string | FlattenInterpolation<ITheme>,
) =>
  `${createMediaQuery(screenSize)} {
    ${css};
  }`;

export const getTransitionValue = (propertyNames: string = 'transform') =>
  `${propertyNames} 0.3s cubic-bezier(0.53, 0.01, 0.36, 1.63);`;

export function getSafeValue<T>(target: unknown, defaultValue: T) {
  return is.notExist(target) ? defaultValue : target;
}

export function getSafeCss(target?: CSSType) {
  return getSafeValue(target, '') as CSSType;
}
