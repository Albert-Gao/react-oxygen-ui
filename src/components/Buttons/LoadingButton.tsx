import React from 'react';
import { DEFAULT_SPACE, getSafeCss } from '../../styles';
import { getRefProp } from '../../utils/utils';
import { BaseBox } from '../Box/BaseBox';
import { IBaseBoxWithTakeRef } from '../Box/BaseBox.type';
import { Image } from '../Image';
import { IButton } from './Button';
import {
  baseNonFlatButtonCSS,
  getBaseButtonCSS,
  INoDefaultHoverCSS,
} from './shared';
import threeDots from './threeDots.svg';

export interface ILoadingButton
  extends IBaseBoxWithTakeRef,
    INoDefaultHoverCSS,
    IButton {
  isLoading?: boolean;
}

export const LoadingButton: React.FC<ILoadingButton> = ({
  isLoading,
  noDefaultHoverCSS,
  children,
  css,
  takeRef,
  ...rest
}) => {
  const cssToAdd = `
    ${isLoading ? 'pointer-events: none' : ''};
    ${getBaseButtonCSS(noDefaultHoverCSS)};
    ${baseNonFlatButtonCSS};
    ${getSafeCss(css)};
  `;

  const a11y = {
    'aria-hidden': 'true',
  };

  return (
    <BaseBox
      as="button"
      position="relative"
      border={0}
      borderColor="transparent"
      borderRadius={0}
      fontSize="bodyHeading"
      px={DEFAULT_SPACE.small}
      minHeight="44px"
      maxHeight="44px"
      bg="primary"
      color="white"
      css={cssToAdd}
      {...getRefProp(takeRef)}
      {...rest}
    >
      <Image
        position="absolute"
        left="50%"
        top="50%"
        src={threeDots}
        alt="loading spinner"
        height="8px"
        opacity={isLoading ? 1 : 0}
        {...a11y}
        transform={
          isLoading
            ? 'translateX(-50%) translateY(-50%) scale(1)'
            : 'translateX(-50%) translateY(-50%) scale(3)'
        }
      />
      <BaseBox
        opacity={isLoading ? 0 : 1}
        transform={isLoading ? 'scale(0.3)' : 'scale(1)'}
      >
        {children}
      </BaseBox>
    </BaseBox>
  );
};
