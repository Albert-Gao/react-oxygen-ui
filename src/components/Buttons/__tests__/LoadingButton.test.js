import 'jest-styled-components';
import 'jest-dom/extend-expect';
import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { LoadingButton } from '../LoadingButton';
import { ThemeProvider } from '../../../styles';

afterEach(cleanup);

describe('Test <LoadingButton />: ', () => {
  test('css prop should be added', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <LoadingButton css={'background-color: black;'} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();

    // the new css is there
    expect(firstChild).toHaveStyleRule('background-color', 'black');

    // the old css has been merged and not lost
    expect(firstChild).toHaveStyleRule('overflow', 'hidden');
  });

  test('undefined css prop should be ignored', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <LoadingButton css={undefined} />
      </ThemeProvider>,
    ).container;

    // invalid css prop will cause `undefined: ;` in result
    // which is invalid
    // so it won't pass this assertion
    expect(firstChild).toMatchSnapshot();
  });

  test('isLoading props works', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <LoadingButton isLoading />
      </ThemeProvider>,
    ).container;

    // invalid css prop will cause `undefined: ;` in result
    // which is invalid
    // so it won't pass this assertion
    expect(firstChild).toMatchSnapshot();

    // the img element
    expect(firstChild.firstChild).toHaveStyleRule(
      'transform',
      'translateX(-50%) translateY(-50%) scale(1)',
    );
  });
});
