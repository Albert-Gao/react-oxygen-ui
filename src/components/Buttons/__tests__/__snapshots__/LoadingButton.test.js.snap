// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <LoadingButton />:  css prop should be added 1`] = `
.c0 {
  box-sizing: border-box;
  padding-left: 32px;
  padding-right: 32px;
  max-height: 44px;
  min-height: 44px;
  font-size: bodyHeading;
  border: 0;
  border-radius: 4px;
  border-color: transparent;
  color: #FFFFFF;
  background-color: #2CAACA;
  background-color: #2CAACA;
  position: relative;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  display: inline-block;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  text-align: center;
  line-height: inherit;
  white-space: nowrap;
  cursor: pointer;
  text-transform: uppercase;
  overflow: hidden;
  background-color: black;
}

.c0:hover {
  opacity: 0.9;
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c0:active {
  -webkit-transform: scale(1) translate(1px,1px);
  -ms-transform: scale(1) translate(1px,1px);
  transform: scale(1) translate(1px,1px);
}

.c1 {
  box-sizing: border-box;
  height: 8px;
  max-width: 100%;
  opacity: 0;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translateX(-50%) translateY(-50%) scale(3);
  -ms-transform: translateX(-50%) translateY(-50%) scale(3);
  transform: translateX(-50%) translateY(-50%) scale(3);
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  object-fit: contain;
}

.c2 {
  box-sizing: border-box;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

<button
  class="c0"
  color="white"
  font-size="bodyHeading"
>
  <img
    alt="loading spinner"
    aria-hidden="true"
    class="c1"
    height="8px"
    opacity="0"
    src="threeDots.svg"
    transform="translateX(-50%) translateY(-50%) scale(3)"
  />
  <div
    class="c2"
    opacity="1"
    transform="scale(1)"
  />
</button>
`;

exports[`Test <LoadingButton />:  isLoading props works 1`] = `
.c0 {
  box-sizing: border-box;
  padding-left: 32px;
  padding-right: 32px;
  max-height: 44px;
  min-height: 44px;
  font-size: bodyHeading;
  border: 0;
  border-radius: 4px;
  border-color: transparent;
  color: #FFFFFF;
  background-color: #2CAACA;
  background-color: #2CAACA;
  position: relative;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  pointer-events: none;
  display: inline-block;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  text-align: center;
  line-height: inherit;
  white-space: nowrap;
  cursor: pointer;
  text-transform: uppercase;
  overflow: hidden;
}

.c0:hover {
  opacity: 0.9;
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c0:active {
  -webkit-transform: scale(1) translate(1px,1px);
  -ms-transform: scale(1) translate(1px,1px);
  transform: scale(1) translate(1px,1px);
}

.c1 {
  box-sizing: border-box;
  height: 8px;
  max-width: 100%;
  opacity: 1;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translateX(-50%) translateY(-50%) scale(1);
  -ms-transform: translateX(-50%) translateY(-50%) scale(1);
  transform: translateX(-50%) translateY(-50%) scale(1);
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  object-fit: contain;
}

.c2 {
  box-sizing: border-box;
  opacity: 0;
  -webkit-transform: scale(0.3);
  -ms-transform: scale(0.3);
  transform: scale(0.3);
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

<button
  class="c0"
  color="white"
  font-size="bodyHeading"
>
  <img
    alt="loading spinner"
    aria-hidden="true"
    class="c1"
    height="8px"
    opacity="1"
    src="threeDots.svg"
    transform="translateX(-50%) translateY(-50%) scale(1)"
  />
  <div
    class="c2"
    opacity="0"
    transform="scale(0.3)"
  />
</button>
`;

exports[`Test <LoadingButton />:  undefined css prop should be ignored 1`] = `
.c1 {
  box-sizing: border-box;
  height: 8px;
  max-width: 100%;
  opacity: 0;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translateX(-50%) translateY(-50%) scale(3);
  -ms-transform: translateX(-50%) translateY(-50%) scale(3);
  transform: translateX(-50%) translateY(-50%) scale(3);
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  object-fit: contain;
}

.c2 {
  box-sizing: border-box;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

.c0 {
  box-sizing: border-box;
  padding-left: 32px;
  padding-right: 32px;
  max-height: 44px;
  min-height: 44px;
  font-size: bodyHeading;
  border: 0;
  border-radius: 4px;
  border-color: transparent;
  color: #FFFFFF;
  background-color: #2CAACA;
  background-color: #2CAACA;
  position: relative;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  display: inline-block;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  text-align: center;
  line-height: inherit;
  white-space: nowrap;
  cursor: pointer;
  text-transform: uppercase;
  overflow: hidden;
}

.c0:hover {
  opacity: 0.9;
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c0:active {
  -webkit-transform: scale(1) translate(1px,1px);
  -ms-transform: scale(1) translate(1px,1px);
  transform: scale(1) translate(1px,1px);
}

<button
  class="c0"
  color="white"
  font-size="bodyHeading"
>
  <img
    alt="loading spinner"
    aria-hidden="true"
    class="c1"
    height="8px"
    opacity="0"
    src="threeDots.svg"
    transform="translateX(-50%) translateY(-50%) scale(3)"
  />
  <div
    class="c2"
    opacity="1"
    transform="scale(1)"
  />
</button>
`;
